driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,134 @@
+        tokens: Vec<Token>,
+        price_source: T,
+        update_interval: Duration,
+    ) -> (Self, JoinHandle<()>) {
+        let price_map = Arc::new(Mutex::new(HashMap::new()));
+        let (sender, receiver) = mpsc::channel::<()>();
+
+        let price_map_ = price_map.clone();
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
+                Ok(()) | Err(RecvTimeoutError::Timeout) => {
+                    // Make sure we don't hold the mutex while this blocking call happens.
+                    match price_source.get_prices(&tokens) {
+                        Ok(prices) => price_map_.lock().unwrap().extend(prices),
+                        Err(err) => log::warn!("price_source::get_prices failed: {}", err),
+                    }
+                }
+                // The owning struct has been dropped.
+                Err(RecvTimeoutError::Disconnected) => break,
+            }
+        });
+
+        // Update the prices immediately.
+        sender.send(()).unwrap();
+        (
+            Self {
+                price_map,
+                _channel_to_thread: sender,
+            },
+            join_handle,
+        )
+    }
+}
+
+impl PriceSource for ThreadedPriceSource {
+    /// Non blocking.
+    /// Infallible.
+    fn get_prices(&self, _tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(self.price_map.lock().unwrap().clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::price_source::MockPriceSource;
+    use super::*;
+    use std::sync::atomic;
+
+    const ORDERING: atomic::Ordering = atomic::Ordering::SeqCst;
+
+    lazy_static::lazy_static! {
+        static ref TOKENS: [Token; 3] = [
+            Token::new(0, "0", 0),
+            Token::new(1, "1", 1),
+            Token::new(2, "2", 2),
+        ];
+    }
+
+    /// Drops tps and joins the handle which ensures that the thread exits and
+    /// hasn't panicked. This is important because the mockall expectation
+    /// panics happen in the thread which would otherwise let the tests pass
+    /// when an expectation is violated.
+    fn join(tps: ThreadedPriceSource, handle: JoinHandle<()>) {
+        std::mem::drop(tps);
+        handle.join().unwrap();
+    }
+
+    #[test]
+    fn thread_exits_when_owner_is_dropped() {
+        let mut ps = MockPriceSource::new();
+        ps.expect_get_prices().returning(|_| Ok(HashMap::new()));
+        let (tps, handle) =
+            ThreadedPriceSource::new(TOKENS.to_vec(), ps, Duration::from_secs(std::u64::MAX));
+        // If the thread didn't exit we would wait forever.
+        join(tps, handle);
+    }
+
+    #[test]
+    fn update_triggered_by_interval() {
+        let mut ps = MockPriceSource::new();
```

Comment:
in test code it's fine, but generally let's please spell out our variable names. It's a few strokes more when you write it but makes comprehension much easier down the line.